microshift.sh

```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+set -exuo pipefail
+
+export LC_ALL=C.UTF-8
+export LANG=C.UTF-8
+
+source tools.sh
+source snc-library.sh
+
+BUNDLE_TYPE="microshift"
+INSTALL_DIR=crc-tmp-install-data
+CRC_VM_NAME=${CRC_VM_NAME:-crc-microshift}
+BASE_DOMAIN=${CRC_BASE_DOMAIN:-testing}
+MIRROR=${MIRROR:-https://mirror.openshift.com/pub/openshift-v4/$ARCH/clients/ocp-dev-preview}
+OPENSHIFT_RELEASE_VERSION=4.12.0-ec.4
+
+if grep -q -i "release 9" /etc/redhat-release
+then
+  echo "Check if system is registered"
+  # Check the subscription status and register if necessary
+  if ! sudo subscription-manager status >& /dev/null ; then
+     sudo subscription-manager register
+  fi
+else
+  echo "This script only works for RHEL-9"
+fi
+
+run_preflight_checks ${BUNDLE_TYPE}
+rm -fr ${INSTALL_DIR} && mkdir ${INSTALL_DIR}
+
+# Download the oc binary for specific OS environment
+download_oc
+OC=./openshift-clients/linux/oc
+
+# Generate a new ssh keypair for this cluster
+# Create a 521bit ECDSA Key
+rm id_ecdsa_crc* || true
+ssh-keygen -t ecdsa -b 521 -N "" -f id_ecdsa_crc -C "core"
+
+# Configure host (RHEL-9) with required internal repo
+sudo tee /etc/yum.repos.d/latest-fdp.repo >/dev/null <<EOF
+[latest-FDP]
+name = latest-FDP-Buildroot-rpms
+baseurl = http://download.eng.bos.redhat.com/rhel-9/nightly/FDP/latest-FDP-9-RHEL-9/compose/Server/\$basearch/os/
+enabled = 1
+gpgcheck = 0
+EOF
+sudo tee /etc/yum.repos.d/latest-rhaos.repo >/dev/null <<EOF
+[latest-RHAOS]
+name = latest-RHAOS-puddle
+baseurl = http://download.lab.bos.redhat.com/rcm-guest/puddles/RHAOS/plashets/4.12-el9/building/\$basearch/os/
+enabled = 1
+gpgcheck = 0
+EOF
+
+rm -fr microshift
+# Clone microshift repo
+git clone https://github.com/openshift/microshift.git
+pushd microshift
+ssh-keygen -t ecdsa -b 521 -N "" -f id_ecdsa_crc -C "core"
+sed -i '/# customizations/,$d' scripts/image-builder/config/blueprint_v0.0.1.toml
+sed -i 's/redhat/core/g' scripts/image-builder/config/kickstart.ks.template
+sed -i 's/epel-release-latest-8.noarch/epel-release-latest-9.noarch/g' scripts/image-builder/configure.sh
+sed -i 's/rhocp-4.11-for-rhel-8-\${BUILD_ARCH}-rpms/latest-RHAOS/g' scripts/image-builder/build.sh
+sed -i 's/fast-datapath-for-rhel-8-\${BUILD_ARCH}-rpms/latest-FDP/g' scripts/image-builder/build.sh
```

Comment:
fwiw, it's a bit confusing that you say rhel9 is mandatory, but microshift from github expects rhel8
Would be useful to document which repo/packages are missing from public repositories so that we can reevaluate at a later time.

Reply:
@cfergeau We want to start creating bundle from RHEL-9 so that we don't have to make any custom changes for M1, as of now `rhocp-4.11-for-rhel-9-$(uname -i)-rpms` and `fast-datapath-for-rhel-9-$(uname -i)-rpms` repo not available even after subscription and we change this with internal repo where dependent packages are available and those internal repo part of this commit. Do you want this detail as part of commit message or as comment?

Reply:
The fact that RHEL9 is a requirement because of M1 support would be useful to have in the commit log or as a comment.

> so that we don't have to make any custom changes for M1

seeing all these `sed`s lines, it seems the tradeoff for this is that we are going to be using custom changes for *all* platforms. In a way it's good because there's wider testing, but if we have rhel9-specific issues, they will impact all our platforms instead of just M1. Will we get help from upstream microshift for rhel9-related problems?

Reply:
As soon as 4.12 RHEL-9 dev preview available we will get required help for rhel-9 related problems. `sed` command around repo is require for RHEL-9 but changing the username to `core` is required for even rhel-8.